network.go

```diff
@@ -0,0 +1,365 @@
+//
+// Copyright (c) 2016 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package virtcontainers
+
+import (
+	"fmt"
+	"math/rand"
+
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/vishvananda/netlink"
+)
+
+// NetworkInterface defines a network interface.
+type NetworkInterface struct {
+	Name   string
+	IPAddr string
+}
+
+// NetworkInterfacePair defines a pair between TAP and virtual network interfaces.
+type NetworkInterfacePair struct {
+	Name             string
+	VirtNetInterface NetworkInterface
+	TAPNetInterface  NetworkInterface
+}
+
+// NetworkConfig is the network configuration related to a network.
+type NetworkConfig struct {
+	PluginName string
+
+	NetworkNS     string
+	NetworkNSPath string
+
+	NetworkPairs []NetworkInterfacePair
+}
```

Comment:
So our `NetworkConfig` should only contain:

* A networking model string ("CNI", "CNM" or "none") e.g. `NetworkModel`
* A networking namespace path. If left empty, this means we need to create one for the container. We could call it `NetNSPath`.

We may need to add a number of interfaces to create, but for now let's assume we want to create only one.

The network pairs is something internal to virtcontainers and should not be exposed. Virtcontainers will build a list of such pairs by scanning the networking namespace or by getting e.g. the CNI plugin information back.

### `NetNSPath`
The first things we need to do related to networking when creating a Pod is to potentially create a networking namespace and have the VM pid join that namespace.
If `NetNSPath` is empty, we will create a persitent networking namespace, i.e. not one linked to the current PID, and then join it. Otherwise we will simply join it.

### `NetworkModel`
This one will let us know what to look for in order to set the netns up:
- "CNI": We will have to look for the default CNI plugin to use, see what [ocicni](https://github.com/rajatchopra/ocicni/blob/master/ocicni.go#L76) does for example. Then:
    1. We call that plugin and the localhost one on the networking namespace we have for the pod
    2. We get the networking information back from CNI (networking setup and interface name)
    3. We create a bridge and a TAP interface, and we bridge the CNI interface and the TAP one through the created bridge.
    4. We add the tap interface information to both our kernel command line and our QEMU parameters
- "CNM": We will need to look for a libnetworking hook. TBD, let's postpone that one.
- "none": We will not set anything related to networking, at all.